### PR TITLE
v4: Convert %(integer: ) to new xlat api

### PR DESCRIPTION
**Pull request by ndptech**

### CODE REVIEW

File: src/lib/unlang/xlat_builtin.c [hunk 1]
```diff
@@ -1117,97 +1117,98 @@ static xlat_action_t xlat_func_explode(TALLOC_CTX *ctx, fr_dcursor_t *out,
...
-		if (vp->vp_length > 8) {
-			break;
-		}
+		if (in_vb->vb_length > 8) goto error;
 
-		if (vp->vp_length > 4) {
-			memcpy(&int64, vp->vp_octets, vp->vp_length);
-			return snprintf(*out, outlen, "%" PRIu64, htonll(int64));
+		if (in_vb->vb_length > 4) {
+			fr_value_box_cast_in_place(ctx, in_vb, FR_TYPE_INT64, NULL);
```
Comment (arr2036) [inline src/lib/unlang/xlat_builtin.c:1178]: That should be UINT64

File: src/lib/unlang/xlat_builtin.c [hunk 2]
```diff
@@ -1117,97 +1117,98 @@ static xlat_action_t xlat_func_explode(TALLOC_CTX *ctx, fr_dcursor_t *out,
...
 
 	uint64_t	int64 = 0;	/* Needs to be initialised to zero */
-	uint32_t	int32 = 0;	/* Needs to be initialised to zero */
-
-	fr_skip_whitespace(fmt);
+	char		buff[40];	/* Temporary string representation special conversions */
 
-	if ((xlat_fmt_get_vp(&vp, request, fmt) < 0) || !vp) return 0;
+	switch (in_vb->type) {
+	case FR_TYPE_UINT8:
```
Comment (arr2036) [inline src/lib/unlang/xlat_builtin.c:1147]: case FR_TYPE_INTEGER:

File: src/lib/unlang/xlat_builtin.c [hunk 3]
```diff
@@ -1117,97 +1117,98 @@ static xlat_action_t xlat_func_explode(TALLOC_CTX *ctx, fr_dcursor_t *out,
...
+	case FR_TYPE_INT32:
+	case FR_TYPE_INT64:
+	case FR_TYPE_TIME_DELTA:
+	case FR_TYPE_SIZE:
+	case FR_TYPE_BOOL:
+		/*
+		 *	Ensure enumeration is NULL so that the integer
+		 *	version of a box is returned
+		 */
+		fr_value_box_cast_in_place(ctx, in_vb, in_vb->type, NULL);
```
Comment (arr2036) [inline src/lib/unlang/xlat_builtin.c:1162]: You just need to set the field in the box to NULL.  If there's issues manipulating the boxes directly add a helper function to clear the enum field in value.c

File: src/lib/unlang/xlat_builtin.c [hunk 4]
```diff
@@ -1117,97 +1117,98 @@ static xlat_action_t xlat_func_explode(TALLOC_CTX *ctx, fr_dcursor_t *out,
...
-
-	case FR_TYPE_UINT32:
-		return snprintf(*out, outlen, "%u", vp->vp_uint32);
-
-	case FR_TYPE_UINT8:
-		return snprintf(*out, outlen, "%u", (unsigned int) vp->vp_uint8);
-
-	case FR_TYPE_UINT16:
-		return snprintf(*out, outlen, "%u", (unsigned int) vp->vp_uint16);
+		snprintf(buff, 40, "%lu", (long unsigned int)ntohl(in_vb->vb_ip.addr.v4.s_addr));
```
Comment (arr2036) [inline src/lib/unlang/xlat_builtin.c:1191]: You can just use the normal cast function here

File: src/lib/unlang/xlat_builtin.c [hunk 5]
```diff
@@ -1117,97 +1117,98 @@ static xlat_action_t xlat_func_explode(TALLOC_CTX *ctx, fr_dcursor_t *out,
...
+		goto output_buff;
 
 	/*
 	 *	Ethernet is weird... It's network related, so it
 	 *	should be bigendian.
 	 */
 	case FR_TYPE_ETHERNET:
-		int64 = vp->vp_ether[0];
+		int64 = in_vb->vb_ether[0];
```
Comment (arr2036) [inline src/lib/unlang/xlat_builtin.c:1199]: I think just cast to a UINT64 and ignore the comment.  If there's not that capability in the box casting function already then add it.

File: src/lib/unlang/xlat_builtin.c (hunk 2 above)
Comment (ndptech) [reply src/lib/unlang/xlat_builtin.c:1147]: FR_TYPE_INTEGER includes FR_TYPE_DATE, which needed different handling to pass tests

**Pull request merged**